compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/inlineclasses/InlineClassAbi.kt

```diff
@@ -0,0 +1,121 @@
+    }
+
+    /**
+     * Get the underlying type of an inline class based on the single argument to its
+     * primary constructor. This is what the current jvm backend does.
+     *
+     * Looking for a backing field does not work for unsigned types, which don't
+     * contain a field.
+     */
+    fun getUnderlyingType(irClass: IrClass): IrType {
+        require(irClass.isInline)
+        val primaryConstructor = irClass.constructors.single { it.isPrimary }
+        return primaryConstructor.valueParameters[0].type
+    }
+
+    /**
+     * Returns a mangled name for a function taking inline class arguments
+     * to avoid clashes between overloaded methods.
+     */
+    fun mangledNameFor(irFunction: IrFunction): Name {
+        val base = when {
+            irFunction is IrConstructor ->
+                "constructor"
+            irFunction.isGetter ->
+                JvmAbi.getterName(irFunction.propertyName.asString())
+            irFunction.isSetter ->
+                JvmAbi.setterName(irFunction.propertyName.asString())
+            irFunction.name.isSpecial ->
+                error("Unhandled special name in mangledNameFor: ${irFunction.name}")
+            else ->
+                irFunction.name.asString()
+        }
+
+        val suffix = when {
+            irFunction.fullValueParameterList.any { it.type.erasedUpperBound.isInline } ->
+                hashSuffix(irFunction)
+            (irFunction.parent as? IrClass)?.isInline == true -> "impl"
+            else -> return irFunction.name
+        }
+
+        return Name.identifier("$base-$suffix")
+    }
+
+    private val IrFunction.propertyName: Name
+        get() = (this as IrSimpleFunction).correspondingPropertySymbol!!.owner.name
+
+    private fun hashSuffix(irFunction: IrFunction) =
+        md5base64(irFunction.fullValueParameterList.joinToString { it.type.eraseToString() })
+
+    private fun IrType.eraseToString() = buildString {
+        append('L')
+        append(erasedUpperBound.fqNameWhenAvailable!!)
+        if (isNullable()) append('?')
+        append(';')
+    }
+}
+
+internal val IrFunction.hasInlineClassParameters: Boolean
+    get() = explicitParameters.any { it.type.erasedUpperBound.isInline } || (this is IrConstructor && constructedClass.isInline)
+
+internal val IrFunction.hasStaticReplacement: Boolean
+    get() = origin != IrDeclarationOrigin.FAKE_OVERRIDE &&
+            (this is IrSimpleFunction || this is IrConstructor && constructedClass.isInline && !isPrimary)
+
+internal val IrFunction.hasMethodReplacement: Boolean
+    get() = dispatchReceiverParameter != null && this is IrSimpleFunction && (parent as? IrClass)?.isInline != true
+
+internal val IrFunction.fullValueParameterList: List<IrValueParameter>
+    get() = listOfNotNull(extensionReceiverParameter) + valueParameters
+
+internal val IrClass.inlineClassFieldName: Name
+    get() = constructors.single { it.isPrimary }.valueParameters.single().name
```

Comment:
There are so many `constructors.single { it.isPrimary }` usages now in this change and elsewhere, that I'd consider adding something like this to the common IR utils:
```
val IrClass.primaryConstructor: IrConstructor?
    get() = constructors.singleOrNull(IrConstructor::isPrimary)
```

Reply:
Done